Site credential login: Prevent redirect callback twice

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

Closes WOOMOB-1545

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

There's a report that nonce-retrieval requests are sometimes duplicated during site credential login.

The cause is that the delegate method `urlSession(_: task: willPerformHTTPRedirection: newRequest:)` is triggered twice: one during the original login request, and another during the manual nonce-retrieval request, which happens occasionally.

This PR fixes the issue by using a separate session for the nonce retrieval request without any delegate. This ensures that the request doesn't accidentally trigger the check for redirection again.

## Testing steps
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

- Uninstall the app. (It seems this happens more frequently on clean installs)
- Enter the address of a self-hosted site.
- Use site credentials to log in to the site.
- Confirm with Proxyman that the nonce-retrieval request (`/wp-admin/admin-ajax.php?action=rest-nonce`) is triggered only once.


## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested and confirmed with simulator iPhone 17. 

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Prevent redirect callback twice

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/257">23.5 ❄️</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adborbas commented:
Hi @itsmeichigo, I could have still reproduced the issue. What should be the expected behavior here? Should we cancel the first request's redirect or not fire the manual?

<img width="1594" height="389" alt="image" src="https://github.com/user-attachments/assets/6b4deb71-2338-4d11-9eeb-4a1993452894" />

itsmeichigo committed:
Simplify logic for site credential login

Remove the manual redirect as cancelling the initial redirect doesn't work properly

itsmeichigo commented:
Thanks @adborbas - I was able to reproduce the issue again and figured the cause might come from our manual redirect logic. It can be too late to cancel the initial redirect, and the current logic seems convoluted so I simplified it by removing the manual redirect in 12269c0a615d2bdff446f2401da806b74226b967.

@hichamboushaba Please take a look if you can - I might have misunderstood your explanation in pe5sF9-1iQ-p2 and made the logic more complicated than it should in #9125. Let me know if there's any potential issue from the change in this PR - I tried the testing steps in that PR and it still works from what I see.

hichamboushaba commented on WooCommerce/Classes/Authentication/SiteCredentialLoginUseCase.swift:
```diff
@@ -172,47 +171,12 @@ private extension SiteCredentialLoginUseCase {
             }
             if let errorMessage = html.findLoginErrorMessage() {
                 throw SiteCredentialLoginError.loginFailed(message: errorMessage)
-            } else {
-                throw SiteCredentialLoginError.invalidLoginResponse
```
Removing this means that any calls to `wp-login.php` that results in `200` status code and that don't match the above conditions will be considered as a success, while it shouldn't.

hichamboushaba reviewed (changes requested):
Thanks @itsmeichigo for your work on this, I like the code simplification, but I think we need to handle some cases to make sure we don't bring any regressions.

Some context here: the goal of disabling the redirection that was suggested in pe5sF9-1iQ-p2 is to make sure we differentiate betwee the `200` status code of `wp-login.php` that should be considered as a failure, and the `200` status code of the nonce request call that should be considered a success.
So the simplification of this PR will cause some issues as pointed in the comment below.

It seems to me that the issue with the previous logic is that using the `willPerformHTTPRedirection` didn't always disable the redirection, I did some tests (with the overload that accepts `completionHandler`, and also with using `URLSessionTaskDelegate` instead of `URLSessionDelegate`) but I couldn't make it stop following the redirect.

But given the goal of ours is to differentiate the error handling depending on the URL, we can achieve the same thing as the P2 plan without disabling the redirect, we can build on top of the changes of this PR to bring back the same logic, please check the attached this [patch](https://github.com/user-attachments/files/23020992/patch.patch), the idea is to check the response's `url` property and then tell if the `200` and `404` was happening for the login URL or the nonce one.

itsmeichigo committed:
Check response of nonce retrieval request

itsmeichigo committed:
Check nonce regex

itsmeichigo committed:
Update the conditional checks for better readability

itsmeichigo commented:
Thanks for the brilliant suggestion @hichamboushaba! I updated the code, this PR is ready for another round 🙏

hichamboushaba commented on WooCommerce/Classes/Authentication/SiteCredentialLoginUseCase.swift:
```diff
@@ -160,10 +160,23 @@ private extension SiteCredentialLoginUseCase {
 
-        switch response.statusCode {
-        case 404:
+        let isNonceUrl = response.url?.absoluteString.hasSuffix(Constants.wporgNoncePath) == true
```
np, do you think adding a comment here just to explain that `isNonceUrl` means the redirect was successful, just to have a link to the class comment that mentions `if the redirect succeeds... login is successful`

itsmeichigo replied:
Sure, done in a175e8640db242f0df5d953f5aedabc6b8ab891d ✅

hichamboushaba reviewed (approved):
Nice work @itsmeichigo, I tested it and it works well.

itsmeichigo committed:
Add comment to clarify the check for nonce URL
